zebrad/src/components/sync/tests/vectors.rs

```diff
@@ -0,0 +1,134 @@
+
+    // This constraint avoids spurious failures after checkpointing has finished
+    assert!(
+        BLOCK_VERIFY_TIMEOUT.as_secs()
+            > 2 * zebra_chain::parameters::NetworkUpgrade::Blossom
+                .target_spacing()
+                .num_seconds() as u64,
+        "Block verify should allow for at least one new block to be generated and distributed"
+    );
+}
+
+/// Test that calls to [`ChainSync::request_genesis`] are rate limited.
+#[test]
+fn request_genesis_is_rate_limited() {
+    // create some counters that will be updated inside async blocks
+    let peer_requests_counter = Arc::new(AtomicU8::new(0));
+    let peer_requests_counter_clone = Arc::clone(&peer_requests_counter);
+    let state_requests_counter = Arc::new(AtomicU8::new(0));
+    let state_requests_counter_clone = Arc::clone(&state_requests_counter);
+    let verifier_requests_counter = Arc::new(AtomicU8::new(0));
+    let verifier_requests_counter_clone = Arc::clone(&verifier_requests_counter);
+
+    let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+    let _guard = runtime.enter();
+
+    // create a fake peer service that respond with `Nil` to `BlocksByHash` or
+    // panic in any other type of request.
+    let peer_service = tower::service_fn(move |request| {
+        match request {
+            zebra_network::Request::BlocksByHash(_) => {
+                // Track the call
+                peer_requests_counter_clone.fetch_add(1, Ordering::SeqCst);
+                // Respond with `Nil`
+                future::ok(zebra_network::Response::Nil)
+            }
+            _ => panic!("no other request is allowed"),
+        }
+    });
+
+    // create a state service that respond with `None` to `Depth` or
+    // panic in any other type of request.
+    let state_service = tower::service_fn(move |request| {
+        match request {
+            zebra_state::Request::Depth(_) => {
+                // Track the call
+                state_requests_counter_clone.fetch_add(1, Ordering::SeqCst);
+                // Respond with `None`
+                future::ok(zebra_state::Response::Depth(None))
+            }
+            _ => panic!("no other request is allowed"),
+        }
+    });
+
+    // create a verifier service that respond always with `GENESIS_PREVIOUS_BLOCK_HASH`
+    let verifier_service = tower::service_fn(move |_| {
+        // Track the call
+        verifier_requests_counter_clone.fetch_add(1, Ordering::SeqCst);
+        // Respond with `GENESIS_PREVIOUS_BLOCK_HASH`
+        future::ok(zebra_chain::parameters::GENESIS_PREVIOUS_BLOCK_HASH)
+    });
+
+    // start the sync
+    let mut chain_sync = ChainSync::new(
+        &ZebradConfig::default(),
+        peer_service,
+        state_service,
+        verifier_service,
+    );
+
+    // run `request_genesis()` with a timeout of 15 seconds
+    runtime.block_on(async move {
+        let _ = timeout(Duration::from_secs(15), chain_sync.request_genesis()).await;
```

Comment:
Will we always have 3 requests here?
What happens if we have requests at 0, 5, 10, and 15 seconds?

Please use a timeout of 13 instead, so it's halfway between the intervals.

Also please get @jvff to show you how to pause and advance the tokio timer.

Reply:
> Also please get @jvff to show you how to pause and advance the tokio timer.

This is now #2270